keps/3258-delayed-admission-check-retries/README.md

```diff
@@ -0,0 +1,545 @@
+   8.1 The message on the QuotaReserved status: The workload has failed admission checks.
+
+At this point, the workload should theoretically be picked up by the scheduler again. However, it's ignored because the admission check remains in a Retry state that is never cleared.
+We mitigated this issue by being more cautious with updates in our custom admission checks. Specifically, by avoiding state changes at the wrong time. However, this behavior was not immediately obvious.
+
+#### Solution: Delayed Processing and State Reconciliation
+
+To address these issues, the implementation will be changed to allow admission check updates after a workload has been evicted. When an admission check state is set to `Retry` or `Ready`, after eviction, Kueue should reset the state to `Pending` and defer any immediate action. Instead of scheduling a timer in `pkg/controller/core/workload_controller.go:807` that sets the requeuing in the future, the controller should rely on the Reconcile function for managing delays. Whenever the workload gets re-submitted to be reconciled, it should search for the latest delay across all admission checks. If the calculated retry time is in the future, the reconciler should reprocess the workload with a delay. If it's in the past, the reconciler should add the workload back into the scheduling rotation.
+
+![This is how the reconciler would work high level](reconciler-flow.png "Reconciler flow")
+
+This diagram shows the updated approach where admission check one sets the state to Retry with a delay of one minute (step 10), and Kueue evicts the workload (step 12 + 13). Crucially, admission check two is able to push its update even after the workload has been evicted, and Kueue will take this new delay into account. The rest of the flow remains the same as before. This approach ensures that all admission checks have the opportunity to set their retry delays, and Kueue will respect the maximum delay across all checks.
+
+```mermaid
+sequenceDiagram
+  autonumber
+  actor user as User
+  participant kueue as Kueue
+  participant kube as Kube API Server
+  participant ac1 as Admission Check 1
+  participant ac2 as Admission Check 2
+
+  user ->> kube: Create Workload
+  kube -->> kueue: New WL event
+  activate kueue
+    kueue -> kueue: Reserve Quota
+    kueue -> kueue: Attach admission check information
+    kueue ->> kube: Push WL update
+  deactivate kueue
+  kube -->> ac1: WL update event
+  activate ac1
+  kube -->> ac2: WL update event
+  activate ac2
+  ac1 ->> ac1: Perform checks
+  ac2 ->> ac2: Perform checks
+  ac2 ->> kube: Set AC to retry with 1m delay
+  deactivate ac2
+  kube -->> kueue: WL update event
+  kueue ->> kube: Reset admission check states
+  kueue ->> kube: Evict Workload
+  activate kueue
+  kueue --> kueue: Don't retry as retry after is not done (in 1m)
+  ac1 ->> kube: Set AC to retry with 2m delay
```

Comment:
When will this AdmissionCheck reset to Pending?

Reply:
Maybe Kueue could reset all AdmissionChecks when assigning quota instead of eviction, wdyt?

Reply:
This sounds like a great idea 🙂 
I'll think more about the potential implications and add it to the doc.